### PR TITLE
Mergeadmin

### PR DESCRIPTION
**Pull request by davis999**
merge code from product admin.

### CODE REVIEW

File: docs/admin_api.md [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+
+* URL: {ip}/products/{productId}
+* method: GET
+* path variable:
+
+| name | type | comments |
+|----|-----|----|
+| productId | String | product id, required |
+
+* response: ProductView
```
Comment (kaka89) [inline docs/admin_api.md:153]: we need an api for fetch product list

File: src/main/java/io/reactivesw/product/application/admin/model/PriceDraft.java [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+import io.reactivesw.model.Money;
+import io.reactivesw.model.Reference;
+
+import lombok.Data;
+import lombok.EqualsAndHashCode;
+
+import java.time.ZonedDateTime;
+
+/**
+ * Created by Davis on 16/11/17.
```
Comment (kaka89) [inline src/main/java/io/reactivesw/product/application/admin/model/PriceDraft.java]: header

File: src/main/java/io/reactivesw/product/application/admin/model/ProductDraft.java [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+import lombok.Data;
+import lombok.EqualsAndHashCode;
+
+import java.util.List;
+
+import javax.validation.constraints.NotNull;
+import javax.validation.constraints.Pattern;
+
+/**
+ * Created by Davis on 16/11/17.
```
Comment (kaka89) [inline src/main/java/io/reactivesw/product/application/admin/model/ProductDraft.java]: header

File: src/main/java/io/reactivesw/product/application/admin/model/ProductVariantDraft.java [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+
+import lombok.Data;
+import lombok.EqualsAndHashCode;
+
+import java.util.List;
+
+import javax.validation.constraints.Pattern;
+
+/**
+ * Created by Davis on 16/11/17.
```
Comment (kaka89) [inline src/main/java/io/reactivesw/product/application/admin/model/ProductVariantDraft.java]: header

File: src/main/java/io/reactivesw/product/application/admin/model/ProductView.java [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+import io.reactivesw.model.Reference;
+import io.reactivesw.product.application.model.ProductCatalogDataView;
+import io.reactivesw.product.application.model.ReviewRatingStatisticsView;
+
+import lombok.Data;
+
+import java.time.ZonedDateTime;
+
+/**
+ * Created by Davis on 16/11/17.
```
Comment (kaka89) [inline src/main/java/io/reactivesw/product/application/admin/model/ProductView.java]: header

File: src/main/java/io/reactivesw/product/application/admin/model/mapper/CategoryOrderHintsMapper.java [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+  private CategoryOrderHintsMapper() {
+  }
+
+  /**
+   * Convert List of CategoryOrderHintView to List of CategoryOrderHint Entity.
+   *
+   * @param models the models
+   * @return the list
+   */
+  public static Set<CategoryOrderHint> toEntity(List<CategoryOrderHintView> models) {
```
Comment (kaka89) [inline src/main/java/io/reactivesw/product/application/admin/model/mapper/CategoryOrderHintsMapper.java]: set or list?

File: src/main/java/io/reactivesw/product/application/admin/model/mapper/CategoryOrderHintsMapper.java [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+    return models.stream().map(
+        model -> {
+          return CategoryOrderHint.build(model.getKey(), model.getOrder());
+        }
```
Comment (kaka89) [inline src/main/java/io/reactivesw/product/application/admin/model/mapper/CategoryOrderHintsMapper.java]: use this??
model -> CategoryOrderHint.build(model.getKey(), model.getOrder());

File: src/main/java/io/reactivesw/product/application/admin/model/mapper/CategoryOrderHintsMapper.java [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+  /**
+   * Convert List of CategoryOrderHint Entity to List of CategoryOrderHintView.
+   *
+   * @param entities the entities
+   * @return the list
+   */
+  public static List<CategoryOrderHintView> toModel(Set<CategoryOrderHint> entities) {
+    return entities.stream().map(
+        entity -> {
+          return new CategoryOrderHintView(entity.getCategoryId(), entity.getOrderHint());
```
Comment (kaka89) [inline src/main/java/io/reactivesw/product/application/admin/model/mapper/CategoryOrderHintsMapper.java]: as said above

File: src/main/java/io/reactivesw/product/application/admin/model/mapper/DimensionsMapper.java [hunk 9]
```diff
@@ -0,0 +1,46 @@
+package io.reactivesw.product.application.admin.model.mapper;
+
+import io.reactivesw.product.application.model.AssetDimensionsView;
+import io.reactivesw.product.domain.model.AssetDimensions;
+
+/**
+ * Created by Davis on 17/2/15.
```
Comment (kaka89) [inline src/main/java/io/reactivesw/product/application/admin/model/mapper/DimensionsMapper.java:7]: header

File: src/main/java/io/reactivesw/product/application/admin/model/mapper/ImageMapper.java [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+  /**
+   * Convert List of Image Entity to List of ImageView.
+   *
+   * @param entities the entities
+   * @return the list
+   */
+  public static List<ImageView> toModel(List<Image> entities) {
+    return entities.stream().map(
+        entity -> {
+          return toModel(entity);
```
Comment (kaka89) [inline src/main/java/io/reactivesw/product/application/admin/model/mapper/ImageMapper.java]: as said above

File: src/main/java/io/reactivesw/product/application/admin/model/mapper/ImageMapper.java [hunk 11]
```diff
@@ -0,0 +1,83 @@
...
+  /**
+   * Convert List of ImageView to List of Image Entity.
+   *
+   * @param models the models
+   * @return the list
+   */
+  public static List<Image> toEntity(List<ImageView> models) {
+    return models.stream().map(
+        model -> {
+          return toEntity(model);
```
Comment (kaka89) [inline src/main/java/io/reactivesw/product/application/admin/model/mapper/ImageMapper.java]: as said above

File: src/main/java/io/reactivesw/product/application/admin/model/mapper/ImageMapper.java [hunk 12]
```diff
@@ -0,0 +1,83 @@
...
+  }
+
+  /**
+   * Convert ImageView to Image Entity.
+   *
+   * @param model the model
+   * @return the image
+   */
+  public static Image toEntity(ImageView model) {
+    Image entity = new Image();
```
Comment (kaka89) [inline src/main/java/io/reactivesw/product/application/admin/model/mapper/ImageMapper.java:72]: should we check null exception? and the same in other mapper class

File: src/main/java/io/reactivesw/product/application/admin/model/mapper/ProductDataMapper.java [hunk 13]
```diff
@@ -0,0 +1,122 @@
...
+  private ProductDataMapper() {
+  }
+
+  /**
+   * Convert ProductDraft to ProductData Entity.
+   *
+   * @param model the model
+   * @return the product data
+   */
+  public static ProductData modelToEntity(ProductDraft model) {
```
Comment (kaka89) [inline src/main/java/io/reactivesw/product/application/admin/model/mapper/ProductDataMapper.java]: toEntity

File: src/main/java/io/reactivesw/product/application/admin/model/mapper/ProductDataMapper.java [hunk 14]
```diff
@@ -0,0 +1,122 @@
...
+    return entity;
+  }
+
+  /**
+   * Convert ProductData Entity to ProductDataView.
+   *
+   * @param entity the entity
+   * @return the product data view
+   */
+  public static ProductDataView entityToModel(ProductData entity) {
```
Comment (kaka89) [inline src/main/java/io/reactivesw/product/application/admin/model/mapper/ProductDataMapper.java]: toModel

File: src/main/java/io/reactivesw/product/application/admin/model/mapper/ProductVariantMapper.java [hunk 15]
```diff
@@ -0,0 +1,114 @@
...
+   * Convert List of ProductVariantDraft to List of ProductVariant Entity.
+   * Set ProductVariant id.
+   *
+   * @param models the models
+   * @return the set
+   */
+  public static Set<ProductVariant> toEntity(List<ProductVariantDraft> models) {
+    Set entities = Sets.newHashSet();
+    for (int i = 0; i < models.size(); i++) {
+      int id = i + 2;
```
Comment (kaka89) [inline src/main/java/io/reactivesw/product/application/admin/model/mapper/ProductVariantMapper.java]: add comments why id start from 2

File: src/main/java/io/reactivesw/product/application/admin/model/mapper/ProductVariantMapper.java [hunk 16]
```diff
@@ -0,0 +1,114 @@
...
+      model.setPrices(PriceMapper.entityToModel(entity.getPrices()));
+    }
+    if (entity.getAttributes() != null) {
+      model.setAttributes(AttributeMapper.toModel(entity.getAttributes()));
+    }
+    if (entity.getImages() != null) {
+      model.setImages(ImageMapper.toModel(entity.getImages()));
+    }
+
+    //availability will be setted in ProductInventoryUtils.mergeInventoryEntryToProduct
```
Comment (kaka89) [inline src/main/java/io/reactivesw/product/application/admin/model/mapper/ProductVariantMapper.java]: delete if useless

File: src/main/java/io/reactivesw/product/application/admin/service/ProductApplication.java [hunk 17]
```diff
@@ -0,0 +1,77 @@
...
+      LOG.debug("can not find product type by id : {}", productTypeId);
+      throw new NotExistException("ProductTypeView Not Found");
+    }
+
+    AttributeConstraintValidator.validate(productType.getAttributes(), productDraft);
+    SkuNameValidator.validate(productDraft);
+
+    ProductView result = productService.createProduct(productDraft);
+
+    LOG.debug("exit.");
```
Comment (kaka89) [inline src/main/java/io/reactivesw/product/application/admin/service/ProductApplication.java]: Log.trace the data, and we should also record who did this action.

File: src/main/java/io/reactivesw/product/application/model/mapper/LocalizedStringMapper.java [hunk 18]
```diff
@@ -47,4 +103,33 @@ private static LocalizedString toModel(Set<LocalizedStringValue> localizedString
...
+  /**
+   * convert LocalizedString  to Set of LocalizedStringValue.
+   *
+   * @param localizedString         LocalizedString
+   * @param localizedStringEntities Set of LocalizedStringValue
+   * @return Set of LocalizedStringValue
+   */
+  private static Set<LocalizedStringValue> toEntity(LocalizedString localizedString,
+                                                    Set<LocalizedStringValue>
+                                                        localizedStringEntities) {
```
Comment (kaka89) [inline src/main/java/io/reactivesw/product/application/model/mapper/LocalizedStringMapper.java]: merge to one line

File: src/main/java/io/reactivesw/product/infrastructure/util/AttributeUtils.java [hunk 19]
```diff
@@ -0,0 +1,39 @@
+package io.reactivesw.product.infrastructure.util;
+
+import io.reactivesw.product.application.model.attribute.AttributeConstraint;
+import io.reactivesw.product.application.model.attribute.AttributeDefinition;
+
+import java.util.List;
+import java.util.stream.Collectors;
+
+/**
+ * Created by Davis on 16/12/20.
```
Comment (kaka89) [inline src/main/java/io/reactivesw/product/infrastructure/util/AttributeUtils.java]: header

File: src/main/java/io/reactivesw/product/infrastructure/util/AttributeUtils.java [hunk 20]
```diff
@@ -0,0 +1,39 @@
...
+  /**
+   * Gets attribute name by constraint.
+   *
+   * @param attributeDefinitions the attribute definitions
+   * @param attributeConstraint  the attribute constraint
+   * @return the attribute name by constraint
+   */
+  public static List<String> getAttributeNameByConstraint(List<AttributeDefinition>
+                                                              attributeDefinitions,
+                                                          AttributeConstraint
```
Comment (kaka89) [inline src/main/java/io/reactivesw/product/infrastructure/util/AttributeUtils.java]: merge to one line

Comment (davis999) [reply src/main/java/io/reactivesw/product/infrastructure/util/AttributeUtils.java]: line length limit.

File: src/main/java/io/reactivesw/product/application/model/mapper/LocalizedStringMapper.java (hunk 18 above)
Comment (davis999) [reply src/main/java/io/reactivesw/product/application/model/mapper/LocalizedStringMapper.java]: line length limit.

File: src/main/java/io/reactivesw/product/application/admin/model/mapper/ProductDataMapper.java (hunk 14 above)
Comment (davis999) [reply src/main/java/io/reactivesw/product/application/admin/model/mapper/ProductDataMapper.java]: fix.

File: src/main/java/io/reactivesw/product/application/admin/model/mapper/ProductDataMapper.java (hunk 13 above)
Comment (davis999) [reply src/main/java/io/reactivesw/product/application/admin/model/mapper/ProductDataMapper.java]: fix.

File: src/main/java/io/reactivesw/product/application/admin/model/mapper/ImageMapper.java (hunk 12 above)
Comment (davis999) [reply src/main/java/io/reactivesw/product/application/admin/model/mapper/ImageMapper.java:72]: I used to check null object when call the method, not check in the method.

File: src/main/java/io/reactivesw/product/application/admin/model/mapper/DimensionsMapper.java (hunk 9 above)
Comment (davis999) [reply src/main/java/io/reactivesw/product/application/admin/model/mapper/DimensionsMapper.java:7]: fix.

File: src/main/java/io/reactivesw/product/application/admin/model/mapper/CategoryOrderHintsMapper.java (hunk 7 above)
Comment (davis999) [reply src/main/java/io/reactivesw/product/application/admin/model/mapper/CategoryOrderHintsMapper.java]: fix.

File: src/main/java/io/reactivesw/product/application/admin/model/mapper/CategoryOrderHintsMapper.java (hunk 6 above)
Comment (davis999) [reply src/main/java/io/reactivesw/product/application/admin/model/mapper/CategoryOrderHintsMapper.java]: fix.

File: src/main/java/io/reactivesw/product/application/admin/model/ProductView.java (hunk 5 above)
Comment (davis999) [reply src/main/java/io/reactivesw/product/application/admin/model/ProductView.java]: fix.

File: src/main/java/io/reactivesw/product/application/admin/model/ProductVariantDraft.java (hunk 4 above)
Comment (davis999) [reply src/main/java/io/reactivesw/product/application/admin/model/ProductVariantDraft.java]: fix.

File: src/main/java/io/reactivesw/product/application/admin/model/ProductDraft.java (hunk 3 above)
Comment (davis999) [reply src/main/java/io/reactivesw/product/application/admin/model/ProductDraft.java]: fix.

File: src/main/java/io/reactivesw/product/application/admin/model/PriceDraft.java (hunk 2 above)
Comment (davis999) [reply src/main/java/io/reactivesw/product/application/admin/model/PriceDraft.java]: fix.

File: docs/admin_api.md (hunk 1 above)
Comment (davis999) [reply docs/admin_api.md:153]: admin api is not complete.

File: src/main/java/io/reactivesw/product/application/admin/model/mapper/CategoryOrderHintsMapper.java (hunk 8 above)
Comment (davis999) [reply src/main/java/io/reactivesw/product/application/admin/model/mapper/CategoryOrderHintsMapper.java]: fix.

File: src/main/java/io/reactivesw/product/application/admin/model/mapper/ImageMapper.java (hunk 10 above)
Comment (davis999) [reply src/main/java/io/reactivesw/product/application/admin/model/mapper/ImageMapper.java]: fix.

File: src/main/java/io/reactivesw/product/application/admin/model/mapper/ImageMapper.java (hunk 11 above)
Comment (davis999) [reply src/main/java/io/reactivesw/product/application/admin/model/mapper/ImageMapper.java]: fix.

File: src/main/java/io/reactivesw/product/application/admin/model/mapper/ProductVariantMapper.java (hunk 15 above)
Comment (davis999) [reply src/main/java/io/reactivesw/product/application/admin/model/mapper/ProductVariantMapper.java]: done.

File: src/main/java/io/reactivesw/product/application/admin/model/mapper/ProductVariantMapper.java (hunk 16 above)
Comment (davis999) [reply src/main/java/io/reactivesw/product/application/admin/model/mapper/ProductVariantMapper.java]: done.

File: src/main/java/io/reactivesw/product/application/admin/service/ProductApplication.java (hunk 17 above)
Comment (davis999) [reply src/main/java/io/reactivesw/product/application/admin/service/ProductApplication.java]: done.

File: src/main/java/io/reactivesw/product/infrastructure/util/AttributeUtils.java (hunk 19 above)
Comment (davis999) [reply src/main/java/io/reactivesw/product/infrastructure/util/AttributeUtils.java]: done.

**Review by kaka89 — APPROVED** [review]

**Pull request merged**